exporter/keys.go

```diff
@@ -235,7 +202,55 @@ func parseKeyArg(keysArgString string) (keys []dbKeyPair, err error) {
 			return keys, fmt.Errorf("couldn't parse db/key string: %s", k)
 		}
 
+		// We want to guarantee at the top level that invalid values
+		// will not fall into the final Redis call.
+		if db == "" && key == "" {
+			log.Debugf("parseKeyArg(): Got empty database and key, pair skipped")
+			continue
+		} else if db == "" || key == "" {
+			log.Errorf("parseKeyArg(): Empty value parsed in pair '%s=%s', skip", db, key)
+			continue
+		}
+
+		number, err := strconv.Atoi(db)
+		if err != nil {
+			return keys, fmt.Errorf("couldn't parse string \"%s\", invalid database index for db \"%s\": %s", k, db, err)
+		}
+		if number < 0 {
+			return keys, fmt.Errorf("db index '%s' should be an non-negative integer value in '%s'", db, k)
+		}
+
 		keys = append(keys, dbKeyPair{db, key})
 	}
 	return keys, err
 }
+
+// scanForKeys returns a list of keys matching `pattern` by using `SCAN`, which is safer for production systems than using `KEYS`.
+// This function was adapted from: https://github.com/reisinger/examples-redigo
+func scanKeys(c redis.Conn, pattern string, count int64) ([]interface{}, error) {
+	iter := 0
+	var keys []interface{}
+
+	if pattern == "" {
```

Comment:
Good you're checking the inputs, do it right away as the first thing before declaring variables.